Use create recursively

pxpm opened the pull request:
## WHY

### BEFORE - What was wrong? What was happening before this PR?

Nested relations were not properly saved nor displayed.

### AFTER - What is happening after this PR?

They save, they display, they work.

## HOW

### How did you achieve that, in technical terms?

This a 100% replacement for https://github.com/Laravel-Backpack/CRUD/pull/4115

Everything is splitted in small commits so we can move them out to new PR's if needed, but they are all related to this PR .. so!

Creating of entries and relations are done recursivelly, so technically any chained relations would work no matter how long the chain, and all of them have the same way of getting direct inputs (parsing) and relations.

Our tests already provide tests for "repeatable inside repeatable", but due to interface issues they are not supported atm just there to show off our relation creations ability. 🥳 


### Is it a breaking change or non-breaking change?

Non breaking in 4.2


### How can we test the before & after?

I've created https://github.com/Laravel-Backpack/demo/pull/348 so this branch can be properly tested, it enables the fake fields and removes the unsuported relations (repeatables inside repeatables).

From my testings:
- Monster works.
- Cave works.
- Story works.
- Fake fields works.
- Tests are passing.

I hope this time is for good after the usual rounds of code review, I think we got this to a pretty working/predictable state.

One of the biggest issues to make all this work together is the fact that we support `dot.notation.relations`. If we didn't support them this PR would be clean as cristal water. But still, I think it's pretty much intuitive. 

Let me know if I let something slip. We talk better about it tomorrow, I want to refactor some variable_names etc etc (in a separate commit).

Pedro

Pedro

pxpm commented on src/app/Library/CrudPanel/Traits/FieldsProtectedMethods.php:
```diff
@@ -144,6 +145,10 @@ protected function makeSureFieldHasEntity($field)
             $field['entity'] = $this->modelMethodHasParameters($model, $possibleMethodName) ? false : $field['name'];
 
+            if ($field['entity']) {
+                $field['nestedEntity'] = $possibleMethodName;
```
This is something I will probably remove tomorrow as I think I endup not needing it.

tabacitu commented on src/app/Library/CrudPanel/Traits/Update.php:
```diff
@@ -113,69 +115,69 @@ private function getModelAttributeValueFromRelationship($model, $field)
                 }
 
-                $related_entry = $related_model->{$relation_method};
+                $related_entry = $related_model->{$relation_method}->withFakes();
```
Hmmm... what is the purpose of adding fakes here? I'm wondering... what if the related model does NOT have the `CrudTrait` on it? Wouldn't this stop it from working?

tabacitu commented on src/app/Library/CrudPanel/Traits/Update.php:
```diff
@@ -113,69 +115,69 @@ private function getModelAttributeValueFromRelationship($model, $field)
                 }
 
-                return $related_entry;
+                return $related_entry->withFakes();
```
Same here... what if the related entry doesn't have a CRUD Panel? Say... it's from a package... like one from Spatie or something.

tabacitu reviewed (changes requested):
Great work here @pxpm ! 🎉🎉🎉

I've given it a first pass, found some naming issues and questions, but not much.

pxpm commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -204,12 +289,31 @@ private function createUpdateOrDeleteOneToOneRelation($relation, $relationMethod
+    /**
+     * Returns the direct inputs parsed for model and relationship creation.
+     *
+     * @param  array  $inputs
```
👍

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -26,21 +21,120 @@ trait Create
+     * @param  mixed  $relationMethod
+     * @return array
+     */
+    private function getFieldsNamesToExclude($fields, $relationMethod)
```
We said we'd rewrite this bit of `getDirectInputsFromInput()` to do something like this:

<img width="1440" alt="Screenshot 2022-01-31 at 18 25 45" src="https://user-images.githubusercontent.com/1032474/151832737-cb595d6a-4291-4493-ba82-55681bfb12c4.png">

Then we can have a method with the right name.

pxpm commented:
@tabacitu refactor done, I guess ... 🙃 

Also added some fixes: 
- For Many relations return collections of items instead an array of items.
- check for `is_a()` of `Support/Collection` and that would work also for `Database/Collection` since the later extends the first. Also added this fix in the fields. (and probably will refactor the relationships fields (select, fetch and fetch_or_create) to include this simple fix instead of all the if and elses, in another PR ...)
- Fixed some fields that we already started the fixing to support collections, but was not broader enough (`checklist, select_multiple and select2_multiple`).

Note for your review tomorrow:
- Heroes should not be tested for now like we talked.
- `select_and_order` does not work (interface, if you create in monster and then manually assign the monster to cave and edit the cave you will see that the values you selected in monster would display properly, but you can't manage the interface, I guess something related with JS)
- dummies are broken, haven't investigated much further.

Don't forget to use https://github.com/Laravel-Backpack/demo/pull/348 to test this, I haven't tried to find a better way of hidding the relations that don't work, so nervermind about it for now, I will refactor that demo branch tomorrow.

tabacitu commented:
Hey @pxpm ,

**Heads-up, do a pull before working on this any further. Details at the bottom.**

Thanks for the update. Let's do one more round please. **These changes you pushed last time break a lot of CRUDs, so I couldn't really test much.**

> For Many relations return collections of items instead an array of items.

🚨 🚨 🚨 🚨 🚨 🚨 🚨  Nope. **Not ok.** This was fixed too far upstream. This will involve editing A LOT of files to account for this change. Including custom field types. And constitutes a HUUUUGE breaking change. **We CANNOT change the type or structure of a field attribute.** It is waaaay to risky, and too big of a breaking change. I bet you did not test this properly after you've made the change - **the demo breaks in all kinds of places now, and I assume it's all because of this**. Please undo this change. No go! 😅  Values are strings, numbers or arrays. Like they've always been. I have another inline comment about this with more of my thoughts about this.

> check for is_a() of Support/Collection and that would work also for Database/Collection since the later extends the first. Also added this fix in the fields. (and probably will refactor the relationships fields (select, fetch and fetch_or_create) to include this simple fix instead of all the if and elses, in another PR ...)

Nice!

> Fixed some fields that we already started the fixing to support collections, but was not broader enough (checklist, select_multiple and select2_multiple).

No! This should have been a HUGE alarm bell. If you make a change, and then need to change a bunch of other files to account for that change... you're doing it wrong. You're pushing a huge breaking change. Values should be arrays. Nothing in the field is a collection right now. When we refactor to collection, we refactor everything to collections. Until then, they're arrays.

----

> Heroes should not be tested for now like we talked.

Ok, done. Removed from sidebar.

> select_and_order does not work (interface, if you create in monster and then manually assign the monster to cave and edit the cave you will see that the values you selected in monster would display properly, but you can't manage the interface, I guess something related with JS)

Ok.

> dummies are broken, haven't investigated much further.

That's not ok. It means we broke something big inside repeatable.

----

I've pushed some changes to this too. I've separated the new methods that deal with inputs into a new `Input` trait. And moved methods arounds. So before you work on this, please do a pull.

pxpm commented on src/app/Library/CrudPanel/Traits/Update.php:
```diff
@@ -112,13 +112,13 @@ private function getModelAttributeValueFromRelationship($model, $field)
             case 'MorphToMany':
                 $relationModels = $relatedModel->{$relationMethod};
-                $result = [];
+                $result = collect();
```
Hey @tabacitu .

This was introduced only in 4.2 so I don't think it breaks anything the way I did it. This is the code in 4.1:

```php
    private function getModelAttributeValue($model, $field)
    {
        if (isset($field['entity'])) {
            $relational_entity = $this->parseRelationFieldNamesFromHtml([$field])[0]['name'];

            $relation_array = explode('.', $relational_entity);

            $relatedModel = $relatedModel = array_reduce(array_splice($relation_array, 0, -1), function ($obj, $method) {
                return $obj->{$method} ? $obj->{$method} : $obj;
            }, $model);

            $relationMethod = Arr::last($relation_array);

            if (method_exists($relatedModel, $relationMethod) && $relatedModel->{$relationMethod}() instanceof HasOne) {
                return $relatedModel->{$relationMethod}->{Arr::last(explode('.', $relational_entity))};
            } else {
                return $relatedModel->{$relationMethod};
            }
        }

        if (is_string($field['name'])) {
            return $model->{$field['name']};
        }

        if (is_array($field['name'])) {
            $result = [];
            foreach ($field['name'] as $key => $value) {
                $result = $model->{$value};
            }

            return $result;
        }
    }
```

So for the `Many` relations it would return an `Database\Collection` with `$relatedModel->{$relationMethod}`, I changed to send a `Support\Collection`, but accounted for that using the `is_a` php function. The reason why we can't directly return the `Database\Collection` is because we need to add the fake fields in the returned results before returning them.

Let me know if I made myself clear here. 

Pedro

tabacitu replied:
Hmm... before this commit I could open the Update page on an Invoice:
![Screenshot 2022-02-01 at 12 52 43](https://user-images.githubusercontent.com/1032474/151955913-03d9230a-7b8a-4a51-8f91-5b84c567df5e.png)

After this commit, I cannot:
![Screenshot 2022-02-01 at 12 53 35](https://user-images.githubusercontent.com/1032474/151955956-fb0d2e6d-11e9-4bde-8ab2-9bd94f5e6a0f.png)

So it's definitely breaking stuff that was previously working (in 4.2).

pxpm replied:
Indeed @tabacitu, sorry when I said breaking previously I meant Breaking Changes. Repeatable can be part of the `Many` group of fields when used as creatable entries, forgot about that one. I will have a look, but it's just apply the same sort to the collection instead of the array. 

This is also new functionality we added in 4.2 so not breaking (BC) nothing here.

tabacitu replied:
No, please don't go modifying yet another file for this to work. Instead, please undo the changes you've made last night about this.

If you want this collection change (which will mean changing x files), please submit a different PR with it towards this branch, explaining why it's needed. We can talk about it there, think of the implications on all fields, analyze if it should be done or not.

The fact is:
- yesterday we had this branch working and 90% of files reviewed;
- today it's not working, and only 60% of the files are reviewed;

So please let's get this branch back to a working state, so we can merge it. That is our no 1 priority. If there are minor issues to be addressed, we can address them in separate PRs.

pxpm commented on src/resources/views/crud/fields/select_grouped.blade.php:
```diff
@@ -32,7 +32,7 @@
                         @foreach ($category->{$field['group_by_relationship_back']} as $subEntry)
                             <option value="{{ $subEntry->getKey() }}"
-                                @if ( ( old($field['name']) && old($field['name']) == $subEntry->getKey() ) || (isset($field['value']) && $subEntry->getKey()==$field['value']))
+                                @if ($current_value === $subEntry->getKey())
```
Indeed, the problem was not here as I was thinking. I think I found the culprit I am looking at it. 
To sum up:
- the `select_grouped` we have in monsters is a fake field and belongsTo relation with `article` that the foreignKey is `select2_from_ajax` in the monsters table. 
```php
     // the relation
    public function article()
    {
        return $this->belongsTo(\Backpack\NewsCRUD\app\Models\Article::class, 'select2_from_ajax');
    }
        // the field
         [                  
                 'label'                      => 'Select_grouped'.backpack_free_badge(),
                 'type'                       => 'select_grouped', //https://github.com/Laravel-Backpack/CRUD/issues/502
                'name'                       => 'select_grouped_id',
                'fake'                       => true,
                'entity'                     => 'article',
                'model'                      => 'Backpack\NewsCRUD\app\Models\Article',
                'attribute'                  => 'title',
                'group_by'                   => 'category', // the relationship to entity you want to use for grouping
                'group_by_attribute'         => 'name', // the attribute on related model, that you want shown
                'group_by_relationship_back' => 'articles', // relationship from related model back to this model
                'tab'                        => 'Selects',
                'wrapperAttributes'          => ['class' => 'form-group col-md-6'],
          ],
```
is this somehow supposed to work ? when you get the `$monster->article` you always get from the `select2_from_ajax` key . I guess having relations without "actually" having them is kind of unexpected. Fake relations ? What are the benefits ? 

What I think makes sense is if you need the same relation twice is because you need to store the relation twice, example you have two selects for articles, one is `current_article` and the other is `sub_article`, you create two relations `article` and `subArticle` and provide the correct keys for each of them. 

Because ... they are relations rigth ? 

So we either consider this a bad implementation in demo and refactor the field definition, or we go down the hole again to support fake relations. 

Let me know,
Pedro

tabacitu replied:
Sorry Pedro, I do not understand the issue. What I understand is that:
- on the 4.2 branch, if you submit the Monster form, it errors and you get back to a form with `select_grouped` and `select2_grouped` pre-filled (old value works);
- on this branch, if you submit the Monster form, it errors and you get back to empty `select_grouped` and `select2_grouped` inputs;

So this change breaks stuff. **What is the purpose of doing this in the first place? Is it just optimization? Or is it fixing an error? How big of an error? How niche of an error?**

It looks to me like we should **remove all changes to `select_grouped` and `select2_grouped` from this branch** - those changes should not have been pushed here. If you want to make changes to these fields, please create a separate PR pointing to this one, explain why the change is needed and we can talk about it there. Talking about it here in this PR is unproductive. This PR has a purpose. This change does not seem to help in that purpose, it is not related.

In the meantime, let's please not let this `select_grouped` proposal for improvement block us from working THE MAIN THING. I feel like this minor issue is holding us back from talking about MUST-DOs. Move this out of the way, please.

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -149,21 +152,36 @@ private function getRelationFieldsWithoutRelationType($relations, $include_neste
+     * @return array
      */
-    private function changeBelongsToNamesFromRelationshipToForeignKey($input)
+    private function changeBelongsToNamesFromRelationshipToForeignKey($input, $belongs_to_fields = [])
```
This function was simple. Now it's complicated. Let's look for a way to make it simple again.

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -149,21 +152,36 @@ private function getRelationFieldsWithoutRelationType($relations, $include_neste
+        if (empty($belongs_to_fields)) {
+            $belongs_to_fields = $this->getFieldsWithRelationType('BelongsTo');
+        } else {
+            foreach ($belongs_to_fields as $field) {
+                if (isset($field['subfields'])) {
+                    $belongs_to_fields = array_merge($field['subfields'], $belongs_to_fields);
+                }
+            }
+            $belongs_to_fields = array_filter($belongs_to_fields, function ($field) {
+                return isset($field['relation_type']) && $field['relation_type'] === 'BelongsTo';
+            });
```
Wait wait wait:
- why are we getting `subfields` merged with `fields`? Won't this have negative effects?
- why are we re-doing the logic of `getFieldsWithRelationType()` here instead of calling `$this->getFieldsWithRelationType('BelongsTo', $fields)`?

I feel like this can be improved quite a bit. Let's talk about it.

tabacitu commented on src/app/Library/CrudPanel/Traits/Create.php:
```diff
@@ -196,20 +178,26 @@ private function createUpdateOrDeleteOneToOneRelation($relation, $relationMethod
 
             // Scenario D
-            if (is_null($relationMethodValue) && $relationDetails['entity'] === $relationMethod) {
+            if (is_null($relationMethodValue) && $relationDetails['relationEntity'] === $relationMethod) {
```
What's the difference between `entity`, `relationEntity` and `baseEntity`?

tabacitu commented on src/resources/views/crud/fields/relationship/entries.blade.php:
```diff
@@ -13,43 +13,22 @@
-    $pivotSelectorField['label'] = $pivotSelectorField['label'] ?? \Str::of($field['name'])->singular()->ucfirst();
-    $pivotSelectorField['relation_type'] = $field['relation_type'];
-    $pivotSelectorField['model'] = $field['model'];
     $pivotSelectorField['ajax'] = $inline_create !== false ? true : ($pivotSelectorField['ajax'] ?? false);
     $pivotSelectorField['data_source'] = $pivotSelectorField['data_source'] ?? ($pivotSelectorField['ajax'] ? url($crud->route.'/fetch/'.$field['entity']) : 'false');
     $pivotSelectorField['minimum_input_length'] = $pivotSelectorField['minimum_input_length'] ?? 2;
     $pivotSelectorField['delay'] = $pivotSelectorField['delay'] ?? 500;
     $pivotSelectorField['placeholder'] = $pivotSelectorField['placeholder'] ?? trans('backpack::crud.select_entry');
```
This probably could use an `array_merge()`

pxpm commented:
@tabacitu make sure you are using https://github.com/Laravel-Backpack/demo/pull/348 and do a `php artisan migrate:fresh --seed` to get a clean database. 

I tested Monsters, Cave, Story and some Pet models. 

- Fake fields don't save on this branch when used as subfields. I've submitted #4129 that points to this branch that adds the support for that.

pxpm commented on src/app/Library/CrudPanel/Traits/FieldsProtectedMethods.php:
```diff
@@ -245,12 +246,34 @@ protected function makeSureSubfieldsHaveNecessaryAttributes($field)
-                $subfield['baseEntity'] = $subfield['baseEntity'] ?? $field['entity'];
+
+                $currentEntity = $subfield['baseEntity'] ?? $field['entity'];
+                // chain the parent field baseEntity if it exists
+                $subfield['baseEntity'] = isset($field['baseEntity']) ? $field['baseEntity'].'.'.$currentEntity : $currentEntity;
```
The relationship entity you already have on entity, the idea of base entity is that you have the chained entity until you reach that point.

Eg: 
Cave -> HasOne -> Monster (entity monster) -> HasOne -> Address (entity monster.address) -> BelongsTo Country in Address (entity monster.address.country)

tabacitu replied:
Ok, got it, thank you! 🙏

pxpm commented on src/app/Library/CrudPanel/Traits/FieldsProtectedMethods.php:
```diff
@@ -245,12 +250,34 @@ protected function makeSureSubfieldsHaveNecessaryAttributes($field)
         }
 
+        if (isset($field['relation_type'])) {
+            switch ($field['relation_type']) {
```
I added a comment here to explain why no other relations are treated.

pxpm commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -61,7 +61,8 @@ public function inferRelationTypeFromRelationship($field)
     public function getOnlyRelationEntity($field)
     {
-        $model = $this->getRelationModel($field['entity'], -1);
+        $entity = isset($field['baseEntity']) ? $field['baseEntity'].'.'.$field['entity'] : $field['entity'];
```
Explained in other comment about the same subject.

tabacitu commented:
EXCELEEEENT!!! Good job @pxpm , I could now test, and can confirm `select_multiple` and `select2_multiple` are working now, as subfields  🎉🎉🎉

- Monster still works 🎉
- Cave works 🎉 (`select_multiple` and `select2_multiple` too)
- Story works 🎉 (`select_multiple` and `select2_multiple` too)

Now let's talk about _how_ it works, or _why_ it works. Let's polish this PR a bit, to get it merged to 4.2. Then we can make separate PRs for the other stuff (like you did for the fake fields 👏👏👏).

tabacitu commented on src/app/Library/CrudPanel/Traits/Update.php:
```diff
@@ -94,103 +90,174 @@ private function getModelAttributeValue($model, $field)
                     }
                 }
 
                 return $result;
```
The fact that we have to convert the collection back to an array ALL multiple field types... it triggers my spidey-sense. Let's see if that's really needed. 

What I understand is that:
- previously, all those fields expected `$field['value']` to be an array;
- now, starting with the 4.2 branch, but on this branch too, `$field['value']` can also be a collection;

**Why? Why did we introduce collection-support on fields, on the 4.2 branch? Is it really _needed_? Are there benefits to it?** If there are no real benefits to it, I'd rather we convert the collection to array upstream (one time), instead of doing in in 5 places in the blade files. 

I believe we can do that here, with something like this:

```suggestion
                // convert the collection back to an array
                if (!empty($result)) {
                    if (is_a($result, \Illuminate\Support\Collection::class)) {
                        $result = $result->pluck($field['attribute'])->toArray();
                    }
                }
                
                return $result;
```

And then remove the support for collections from the fields themselves.

**Am I wrong? Can we do that @pxpm ?** Let me know if you want me to jump on a call, if it's easier to explain that way.

tabacitu replied:
I'm asking because... fields are supposed to be something _simple to create_. 

Custom multiple inputs shouldn't have to treat a bunch of cases themselves (is array, is collection, is db collection, is string), they should know "_ok, value is this format, I parse it, and make sure the input is in the right format_".

Maybe we've offset this complexity to the fields, when we should have treated it in the Update trait.

pxpm replied:
Solved

tabacitu commented on src/resources/views/crud/fields/relationship.blade.php:
```diff
@@ -35,7 +35,8 @@
             // TODO: if relationship has `isOneOfMany` on it, load a readonly select; this covers:
```
Is this TODO done now? Or do we still need to do something here?

tabacitu replied:
```suggestion
            // if relationship has `isOneOfMany` on it, load a readonly select; this covers:
```

tabacitu commented on src/app/Library/CrudPanel/Traits/Input.php:
```diff
@@ -0,0 +1,195 @@
+     * @param  mixed  $relationMethod
+     * @return array The formatted relation details.
+     */
+    private function getRelationDetailsFromInput($input, $crudFields = [], $relationMethod = false)
```
I've been trying to understand the structure of `$relationDetails` (finally!) and I think I'm close to understanding it. The name certainly hasn't helped, I think a more appropriate name would be `$relationTree` or `$relationHierarchy` - right?

AFAIK, the difference between this and fields is that this is nested - each relationship inside its parent.

Let's have a quick debugging session, to understand its structure. I feel we can achieve more in a call. Maybe we change its structure now, maybe we leave it like it is... but right now it seems a bit disorganized, I think one-two changes will greatly improve its readability.

Thanks!

tabacitu commented:
Ok so I've given this a hard look and (good news!) we're 10/17 files approved. Please take a look at [my comment here](https://github.com/Laravel-Backpack/CRUD/pull/4127/files#r797336307), if something like that is possible, it should resolve 5 more files 🙀

Then for the 2 remaining files, I think we can have a quick talk, see what (if anything) we do about my concerns.

Overall... I feel like we're very _very_ close to having this merged.

tabacitu commented:
![Screenshot 2022-02-02 at 17 58 02](https://user-images.githubusercontent.com/1032474/152189806-8592a6f6-d31e-400e-9510-a9cc7be8690b.png)

OMGOMGOMGOMGOMGOM!!!!
